gaps/gap-16_golem_compose/gap-16_golem_compose.md

```diff
@@ -0,0 +1,177 @@
+---
+gap: GAP-16
+title: Multi-service application deployment framework for Golem Network
+description: A solution for defining, deploying and managing apps based on multiple services operating within one or more Golem VPNs
+author: Jakub Mazurek (@zakaprov) <jakub.mazurek@golem.network>
+status: Draft
+type: Feature
+requires (*optional): <GAP number(s)>
+replaces (*optional): <GAP number(s)>
+---
+
+## Abstract
+This GAP introduces a proposal for a language-agnostic layer of abstraction over the Golem VPN and its high-level APIs (`yapapi`, `yajsapi`).
+The main element of this proposal is the introduction of a **network configuration descriptor** file which can be used to describe a deployment of a set of services to be run within the Golem Network.
+This configuration descriptor is intended to be used together with high-level APIs to enable deploying and managing the nodes described within the descriptor.
+
+## Motivation
+### Why?
+The main motivation for this proposal is providing support for development of distributed applications (dApps) deployed within the Golem Network.
+### Goals
+- enable reusable application deployments through a version-control-friendly configuration format (descriptor files)
+- simplify creating complex network topologies (adding a layer of abstraction over the existing Golem VPN functionality)
+- allow for easily starting and tearing down multiple different payloads within the Golem Network
+
+## Specification
+### Configuration descriptor file
+Here's an example of such a descriptor file:
+```
+version: "0.0.1"
+
+auth:
+  address: "http://rest-api:6000"
+  app_key: "$YAGNA_APP_KEY"
+  gsb_url: "tcp://gsb-url:6666"
+
+payment:
+  budget: 10  # GLM
+  driver: "polygon"
+  network: "mainnet"
+
+payloads:
+  nginx:
+    runtime: "vm" # runtimes should have their own capabilities assigned
+    params:
+      image: "image-hash"
+      repo-url: "http://girepo.dev.golem.network:8000"
+    constraints:
+      "golem.inf.cpu.cores": 2
+      "golem.inf.mem.gib": 4
+      "golem.inf.storage.gib": 10
+    capabilities:
+      - "vpn"
+
+networks:
+  default:
+    ip: "192.168.0.0/24"
+
+nodes:
+  reverse-proxy:
+    payload: "nginx"
+    args:
+      deploy: # dict, optional
+      start: # list, optional
+    networks:
+      - "name": "default"
+        "address": "192.168.0.2"  # optional
+    environment:
+      - "SOME_VARIABLE=1234"
+```
+
+[TODO: detailed description of each field] 
```

Comment:
I wouldn't worry about detailed descriptions of each field at this point. However it would be useful to list all "sections" of the descriptor (ie. are `auth`, `payment`, `payloads`, `network` and `nodes` all the sections required at current state of Golem?)